### ISSUE TITLE
[apollo-link-error] TS2416: Property 'request' in type 'ErrorLink' is not assignable to the same property in base type 'ApolloLink'

### ISSUE THREAD

**Issue by rosskevin** [apollographql/apollo-link#599]
**Expected Behavior**
Transpiles with latest packages.

**Actual Behavior**
Upgraded to latest packages:
```
 apollo-link          ^1.2.1  →  ^1.2.2 
 apollo-link-context  ^1.0.7  →  ^1.0.8 
 apollo-link-error    ^1.0.7  →  ^1.0.8 
```

Using typescript 2.8.1:

```
$ tsc -d --project tsconfig.build.json
../../node_modules/apollo-link-error/lib/index.d.ts(22,5): error TS2416: Property 'request' in type 'ErrorLink' is not assignable to the same property in base type 'ApolloLink'.
  Type '(operation: any, forward: any) => Observable<ExecutionResult>' is not assignable to type '(operation: Operation, forward?: NextLink) => Observable<FetchResult<Record<string, any>, Record<...'.
    Type 'Observable<ExecutionResult>' is not assignable to type 'Observable<FetchResult<Record<string, any>, Record<string, any>>>'.
      Type 'ExecutionResult' is not assignable to type 'FetchResult<Record<string, any>, Record<string, any>>'.
        Type 'ExecutionResult' is not assignable to type 'ExecutionResult'. Two different types with this name exist, but they are unrelated.
          Types of property 'errors' are incompatible.
            Type 'ReadonlyArray<GraphQLError>' is not assignable to type 'GraphQLError[]'.
              Property 'push' is missing in type 'ReadonlyArray<GraphQLError>'.
```

**Issue Labels**

- [ ] has-reproduction
- [ ] feature
- [ ] docs
- [x] blocking
- [ ] good first issue

### TL;DR

It appears to be a signature mismatch on `request`.  Whereas the base class `ApolloLink` specifies a fully typed signature, the subclass in the apollo-link-error remains silent on arg types and has a different return type:
- base class https://github.com/apollographql/apollo-link/blob/83473cdcd3d7e7a2563a44f3138414ebb8d82583/packages/apollo-link/src/link.ts#L114-L117
- sub class https://github.com/apollographql/apollo-link/blob/83473cdcd3d7e7a2563a44f3138414ebb8d82583/packages/apollo-link-error/src/index.ts#L69

**Comment by evans** [comment]
@rosskevin Thank you for flagging the issue! I'll look at it asap and have a new version of apollo-link-error published asap. I'd like to have an end to end test put together to make sure we capture this sort of thing in the futrue

**Comment by evans** [comment]
It looks like it's the different return type that causes that issue. I'm not able to recreate it locally. Maybe because of different graphql type versions. What version are you using?

Regardless, we'll need to change up the return type for apollo-link-error

**Comment by evans** [comment]
@rosskevin this should be fixed in apollo-link-error@1.0.9 Let me know if you run into issues

**Comment by rosskevin** [comment]
@evans - 1.0.9 confirmed fixed.  Thank you!
